feat(linter/plugins): rule options

lilnasy opened the pull request:
- Implements #14825
- Rebased #15822. Refactored. Pruned the slop. Tests passing.

### Big changes from the original approach
- [x] Options IDs are stored in their own arrays. This was laid out in #16158. The original approach was to store `ruleId`-`optionId` pairs as tuples in an array - several heap allocated objects and excessive pointer indirection.
- [x] The rust code populates the options by calling a new JS callback named `setupConfigs()`. I expect we would use this callback for more plugin related data, including settings which are currently implemented in a hacky way.
- [ ] Investigate a refactor to avoid requiring a mutable reference to `ExternalPluginStore` everywhere.

<details>
<summary>

#### Smaller Changes

</summary>

- Removed the extra fixture ("working version"). There were no meaningful changes compared to the first fixture. Probably slop.

</details>

wagenet committed:
feat(linter/plugins): rule options

overlookmotel commented:
> Refactor to avoid requiring a mutable reference to `ExternalPluginStore` everywhere.

I know it was met that said this should be possible, but I could very well be wrong. If it's not possible, don't worry about it.

lilnasy commented on apps/oxlint/src/lint.rs:
```diff
@@ -291,6 +291,18 @@ impl CliRunner {
+                );
+                return CliRunResult::InvalidOptionConfig;
+            }
+        }
```
I could use help here. I exposed `setup_configs()` for the `oxlint` crate by putting it on `ExternalPluginStore` only because it's the first place I found. I could do with a few pointers on separation of concerns regarding these structs.

overlookmotel replied:
Yeah, the types are a bit of a convoluted mess here. I'm tied up today. Will take a look tomorrow.

lilnasy replied:
- Guessing this is related to https://github.com/oxc-project/oxc/issues/12437?

overlookmotel replied:
I think the way you've done it is good!

Personally I find all the logic around config in linter labyrinthine and confusing/confused. You're asking about separation of concerns. The problem IMO is that there isn't any! Or at least if there is, it's not clear what it is (to me anyway).

overlookmotel commented on apps/oxlint/src/run.rs:
```diff
@@ -40,12 +40,28 @@ pub type JsLintFileCb = ThreadsafeFunction<
+    // Return value
+    String, // Result ("ok" or error message)
+    // Arguments (repeated)
+    FnArgs<(String,)>,
```
Because there's only 1 argument here, can skip the `FnArgs` abstraction.

```suggestion
pub type JsSetupConfigsCb = ThreadsafeFunction<
    // Arguments
    String, // Options array as JSON
    // Return value
    String, // Result ("ok" or error message)
    // Arguments (repeated)
    String,
```

If that gives you problems (weird errors elsewhere due to the change), don't worry about it - we can change it in a follow-up. NAPI-RS's API is pretty hairy.

overlookmotel commented on apps/oxlint/test/fixtures/custom_plugin_with_options/plugin.ts:
```diff
@@ -0,0 +1,42 @@
+              });
+            }
+          },
+        };
```
Would be simpler just to dump `context.options` to the snapshot as JSON, and return an empty visitor `{}`.

overlookmotel commented on crates/oxc_linter/src/external_plugin_store.rs:
```diff
@@ -15,13 +17,18 @@ define_index_type! {
     plugins: IndexVec<ExternalPluginId, ExternalPlugin>,
     plugin_names: FxHashMap<String, ExternalPluginId>,
     rules: IndexVec<ExternalRuleId, ExternalRule>,
+    options: IndexVec<ExternalOptionsId, serde_json::Value>,
```
This would be a bit more type-safe - ensure that each entry is a `Vec` (so an array on JS side):

```suggestion
    options: IndexVec<ExternalOptionsId, Vec<serde_json::Value>>,
```

overlookmotel commented on crates/oxc_linter/src/external_plugin_store.rs:
```diff
@@ -119,6 +129,27 @@ impl ExternalPluginStore {
+        }
+
+        self.options.push(options)
+    }
```
As far as I can see, `ESLintRule::config` field is always `None` or it's `Some` containing `serde_json::Array` with not-zero length. It's never `Some` containing any other type.

https://github.com/oxc-project/oxc/blob/43a6c3209e7c03e3b83d37879ddf7ff919f44d2b/crates/oxc_linter/src/config/rules.rs#L269-L306

This is poor design. `ESLintRule::config` should be `Option<Vec<serde_json::Value>>`, not `Option<serde_json::Value>`. But that's another story...

For now, here could just do:

```rs
let options = options.as_array().unwrap();
debug_assert!(!options.is_empty(), "`options` should never be an empty `Vec`");
self.options.push(options)
```

overlookmotel reviewed (commented):
I'll need to go through this again in depth. But looking good!

Some small points below. You seem to be picking up Rust much faster than I did!

overlookmotel commented on apps/oxlint/src-js/plugins/config.ts:
```diff
@@ -0,0 +1,26 @@
+ * @param optionsJSON - JSON serialization of an array containing all rule options across all configurations.
+ * @returns "ok" on success, or error message on failure
+ */
+export function setupConfigs(optionsJSON: string): string {
```
Your comment is not helpful, Mr Robot!

lilnasy commented on apps/oxlint/test/fixtures/custom_plugin_with_options/plugin.ts:
```diff
@@ -0,0 +1,42 @@
+          wrongValue: "Expected value to be {{expected}}, got {{actual}}",
+          noOptions: "No options provided",
+        },
+      },
```
Yeah, good catch.

overlookmotel committed:
docs: update comments

overlookmotel committed:
style: formatting

overlookmotel committed:
style: use `for ... of` loop

This code is debug-only, so perf doesn't matter.

overlookmotel committed:
docs: shorten comment

overlookmotel committed:
refactor: export `ExternalLinterSetupConfigsCb` from `oxc_linter` crate

overlookmotel committed:
docs: revise comment

overlookmotel committed:
docs: revise comment

overlookmotel committed:
move calling `setupConfigs` to after check for `list_rules`

overlookmotel committed:
docs: add comment

overlookmotel committed:
test: combine test into existing test fixture

overlookmotel committed:
test: test all JSON value types

overlookmotel committed:
style: shorten code (dereference early)

overlookmotel committed:
refactor: convert to `PathBuf` explicitly

Personally I try to avoid `.into()` as it's not clear from the code "into what?". `T::from` is equivalent, but more explicit.

overlookmotel committed:
style: longer var names

overlookmotel committed:
test: fix tests

Make sure severity is as expected. `is_warn_deny()` returns true for either `Warn` or `Deny`. We want to be more specific.

overlookmotel committed:
style: combine imports

overlookmotel committed:
test: make test more realistic

overlookmotel committed:
refactor: shorten code

Dereference during destructuring.

overlookmotel committed:
style: rename vars

Unlike in JS, in Rust you can have multiple bindings in same scope with same name. The old binding is destroyed when the new one is created. This is quite idiomatic in Rust.

overlookmotel committed:
style: full stops on comments

overlookmotel committed:
refactor: remove unnecessary type annotation

overlookmotel committed:
refactor: assume `options` passed to `add_options` is an array

overlookmotel reviewed (commented):
I've pushed some changes. All nits and code style apart from these ones which are more substantive:

* 330604e980ea3135875dbeb26ae12fa20839fd9d
* 63881d71c052a3a3ed96382166f1d1376c5287f7
* 8ecc0714244792d50ae00f782f967c7579e89bd3
* 8f2a6e282b2cb2b29170493ae44f972840b616e5
* 736841bf77f28ce4bd6da2b146083cd9c7ad0410

Even those ones are all still minor though.

Great work! This will be in today's release.

overlookmotel commented on apps/oxlint/src-js/plugins/options.ts:
```diff
@@ -22,8 +22,10 @@ export type Options = JsonValue[];
+// All rule options.
+// Indexed by options ID sent alongside ruleId for each file.
+// Element 0 is always the default options (empty array).
+export let allOptions: Readonly<Options>[] = [DEFAULT_OPTIONS];
```
#16358
